File: Data Structures & Algorithms/two-integer-sum/submission-8.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        values = {}
        for i in range(len(nums)):
            comp = target - nums[i]
            if comp in values:
                return [values[comp], i]
            values[nums[i]] = i
        return []

File: Data Structures & Algorithms/combination-target-sum/submission-8.cpp
class Solution {
public:
    vector<vector<int>> res;
    void backtrack(vector<int>& nums, int target, vector<int>& cur, int idx){
        if(target == 0){
            res.push_back(cur);
            return;
        }

        for(int i = idx; i < nums.size(); i++){
            if(target - nums[i] < 0)
                continue;
            
            cur.push_back(nums[i]);
            backtrack(nums, target - nums[i], cur, i);
            cur.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> cur;
        backtrack(nums, target, cur, 0);
        return res;
    }
};
